crates/walrus-service/src/event/event_processor/coordination.rs

```diff
@@ -82,30 +85,38 @@ impl CatchupCoordinationState {
 
     /// Mark that checkpoint tailing has started.
     pub fn mark_tailing_started(&self) {
+        tracing::debug!("marking tailing started");
         self.is_tailing_stopped
             .store(false, std::sync::atomic::Ordering::Release);
     }
 
     /// Notify that checkpoint tailing has fully stopped and wake any waiters.
     pub fn notify_tailing_stopped(&self) {
+        tracing::debug!("notifying tailing stopped");
         self.is_tailing_stopped
             .store(true, std::sync::atomic::Ordering::Release);
         self.tailing_stopped_notify.notify_one();
     }
 
+    /// Check if checkpoint tailing is currently stopped.
+    pub fn is_tailing_stopped(&self) -> bool {
+        self.is_tailing_stopped.load(Ordering::Acquire)
+    }
+
     /// Wait until checkpoint tailing is stopped, with timeout.
-    /// Returns true if stopped, false on timeout.
+    /// Returns true if stopped, false on timeout. `is_tailing_stopped`
+    /// is the source of truth for the tailing stop state. A stale spurious
+    /// notification might cause an extra wake up  but it will be ignored by the loop.
     pub async fn wait_for_tailing_stopped(&self, timeout: std::time::Duration) -> bool {
-        if self
-            .is_tailing_stopped
-            .load(std::sync::atomic::Ordering::Acquire)
-        {
-            return true;
-        }
-        match tokio::time::timeout(timeout, self.tailing_stopped_notify.notified()).await {
-            Ok(()) => true,
-            Err(_) => false,
-        }
+        let fut = async {
+            loop {
+                if self.is_tailing_stopped.load(Ordering::Acquire) {
+                    return true;
+                }
+                self.tailing_stopped_notify.notified().await;
```

Comment:
if `is_tailing_stopped == false`, and there is already a `tailing_stopped_notify` stored from the node start up, isn't the case that `self.tailing_stopped_notify.notified().await;` will return immediately, but is_tailing_stopped is still false?

Reply:
done

Reply:
Thanks Sadhan!

Just one last suggestion: create `future = self.tailing_stopped_notify.notified()` before checking `is_tailing_stopped`, and poll the future if `is_tailing_stopped` is false.

In the current code, it's possible that `notify_waiters()` is called in between checking `is_tailing_stopped` and `self.tailing_stopped_notify.notified().await;`. When that happens, `self.tailing_stopped_notify.notified().await` will never get wake up.

This is very very unlikely, but I'd prefer let's not leave opportunities for race conditions.

Reply:
Great suggestion @halfprice!